### PR TITLE
Code cleanup

### PR DESCRIPTION
**Pull request by testn**

### CODE REVIEW

**Comment by juherr** [comment]
Good luck, @cbeust doesn't like this kind of huge diff. I've already tried too ;)
You should cut it into multiple pr or at least multiple commits.
To be honest, I didn't have the courage to try to read the whole of the pr.

**Comment by cbeust** [comment]
Yes. @testn I appreciate all the work you're doing but it would be much easier to review these if they're broken down in small, focused individual PR's.

**Comment by testn** [comment]
the delta itself is quite small and independence. You can look at them individually. There are just too many changes to split into individual PRs

File: src/main/java/org/testng/xml/XmlTest.java [hunk 1]
```diff
@@ -582,21 +585,18 @@ public String toXml(String indent) {
   @Override
   public String toString() {
 //    return toXml("");
-    StringBuffer result = new StringBuffer("[Test: \"" + m_name + "\"")
-      .append(" verbose:" + m_verbose);
+    StringBuilder result = new StringBuilder("[Test: \"" + m_name + "\"").append(" verbose:").append(m_verbose);
```
Comment (juherr) [inline src/main/java/org/testng/xml/XmlTest.java]: You missed `new StringBuilder().append("[Test: \"").append(m_name).append("\"")`

File: src/main/java/org/testng/xml/XmlTest.java [hunk 2]
```diff
@@ -360,8 +360,8 @@ public String getParameter(String name) {
     for (Map.Entry<String, String> parameter : parameters.entrySet()) {
       result.put(parameter.getKey(), parameter.getValue());
     }
-    for (String key : m_parameters.keySet()) {
-      result.put(key, m_parameters.get(key));
+    for (Map.Entry<String, String> entry  : m_parameters.entrySet()) {
```
Comment (juherr) [inline src/main/java/org/testng/xml/XmlTest.java]: Replacing the 2 loops by 

```java
result.putAll(getSuite().getParameters());
result.putAll(m_parameters);
```

seems a better approach for me.

File: src/main/java/org/testng/xml/XmlClass.java [hunk 3]
```diff
@@ -288,8 +288,8 @@ public void setParameters(Map<String, String> parameters) {
     for (Map.Entry<String, String> parameter : parameters.entrySet()) {
       result.put(parameter.getKey(), parameter.getValue());
     }
-    for (String key : m_parameters.keySet()) {
-      result.put(key, m_parameters.get(key));
+    for (Map.Entry<String, String> entry  : m_parameters.entrySet()) {
```
Comment (juherr) [inline src/main/java/org/testng/xml/XmlClass.java]: Like below: putAll instead of loop.

File: src/main/java/org/testng/reporters/JUnitReportReporter.java [hunk 4]
```diff
@@ -81,12 +82,14 @@ public void generateReport(List<XmlSuite> xmlSuites, List<ISuite> suites,
...
 //    }
 
+    SimpleDateFormat df = new SimpleDateFormat("dd MMM yyyy HH:mm:ss");
     for (Map.Entry<Class<?>, Set<ITestResult>> entry : results.entrySet()) {
       Class<?> cls = entry.getKey();
       Properties p1 = new Properties();
       p1.setProperty("name", cls.getName());
       Date timeStamp = Calendar.getInstance().getTime();
-      p1.setProperty(XMLConstants.ATTR_TIMESTAMP, timeStamp.toGMTString());
+      String timestampStr = df.format(timeStamp) + " GMT";
```
Comment (juherr) [inline src/main/java/org/testng/reporters/JUnitReportReporter.java]: Why do you add GMT by hands? z attribute should be used instead.
BTW, I wonder if `DateFormat.getDateInstance().format(myDate)` would not be a better approach.

File: src/main/java/org/testng/remote/strprotocol/GenericMessage.java [hunk 5]
```diff
@@ -37,25 +37,6 @@ public void setTestCount(int testCount) {
...
-//    m_properties = props;
-//  }
-
-//  public GenericMessage addProperty(final String propName, final Object propValue) {
-//    m_properties.put(propName, propValue);
-//
-//    return this;
-//  }
-
-  public GenericMessage addProperty(final String propName, final int propValue) {
```
Comment (juherr) [inline src/main/java/org/testng/remote/strprotocol/GenericMessage.java]: Why do you remove this method? It could be used by testng extension like ide plugins.

File: src/main/java/org/testng/internal/Utils.java [hunk 6]
```diff
@@ -748,23 +748,20 @@ public static String replaceSpecialCharacters(String fileName) {
...
 
       //For Overwrite the file.
       to.getParentFile().mkdirs();
-      OutputStream out = new FileOutputStream(to);
-
-      byte[] buf = new byte[1024];
-      int len;
-      while ((len = in.read(buf)) > 0){
-        out.write(buf, 0, len);
+      try (OutputStream out = new FileOutputStream(to)) {
```
Comment (juherr) [inline src/main/java/org/testng/internal/Utils.java]: Why not using the same try-with-resources block for in and out?

File: src/main/java/org/testng/internal/TestNGClassFinder.java [hunk 7]
```diff
@@ -203,10 +203,9 @@ public TestNGClassFinder(ClassInfoMap cim,
     //
     // Add all the instances we found to their respective IClasses
     //
-    for(Class c : m_instanceMap.keySet()) {
-      List<Object> instances= m_instanceMap.get(c);
-      for(Object instance : instances) {
-        IClass ic= getIClass(c);
+    for(Map.Entry<Class, List<Object>> entry : m_instanceMap.entrySet()) {
+      for(Object instance : entry.getValue()) {
+        IClass ic= getIClass(entry.getKey());
```
Comment (juherr) [inline src/main/java/org/testng/internal/TestNGClassFinder.java]: Extract the line from the second loop. It will avoid to look up the same thing many times.

File: src/main/java/org/testng/internal/MethodHelper.java [hunk 8]
```diff
@@ -351,7 +352,11 @@ public boolean hasNext() {
 
   @Override
   public Object[] next() {
-    return m_objects[m_count++];
+    if (m_count < m_objects.length) {
```
Comment (juherr) [inline src/main/java/org/testng/internal/MethodHelper.java]: It's a personal taste, but I prefer:

```java
if (m_count >= m_objects.length) {
  throw new NoSuchElementException();
}
return m_objects[m_count++];
```

File: src/main/java/org/testng/reporters/JUnitReportReporter.java (hunk 4 above)
Comment (testn) [inline src/main/java/org/testng/reporters/JUnitReportReporter.java]: You're right... the correct format should actually be 
```
    SimpleDateFormat df = new SimpleDateFormat("d MMM yyyy HH:mm:ss z");
    df.setTimeZone(TimeZone.getTimeZone("GMT"));
```
`DateFormat.getDateInstance()` is not good enough as toGMTString() returns a specific date string format.

File: src/main/java/org/testng/remote/strprotocol/GenericMessage.java (hunk 5 above)
Comment (testn) [inline src/main/java/org/testng/remote/strprotocol/GenericMessage.java]: If it is used, it might not be working as the method is implemented as an infinite recursive after other implementations have been removed.

File: src/main/java/org/testng/TestRunner.java [hunk 9]
```diff
@@ -1483,7 +1461,7 @@ public static int getVerbose() {
     return m_verbose;
   }
 
-  public void setVerbose(int n) {
+  public static void setVerbose(int n) {
```
Comment (juherr) [inline src/main/java/org/testng/TestRunner.java]: As each <test> node may have its own verbose value, `m_verbose` has to be non-static.

File: src/main/java/org/testng/TestNG.java [hunk 10]
```diff
@@ -875,9 +875,7 @@ private void initializeCommandLineSuitesGroups() {
     }
   }
   private void addReporter(Class<? extends IReporter> r) {
-    if (! m_reporters.contains(r)) {
-      m_reporters.add(ClassHelper.newInstance(r));
-    }
+    m_reporters.add(ClassHelper.newInstance(r));
```
Comment (juherr) [inline src/main/java/org/testng/TestNG.java]: It doesn't sound good: you will create a new instance of r, even if r already exists.

**Comment by juherr** [comment]
I took time to review the PR. I'm not sure it will be easy to follow next changes now.

File: src/main/java/org/testng/TestNG.java (hunk 10 above)
Comment (testn) [inline src/main/java/org/testng/TestNG.java]: `m_reporters.contains(r)` always return false since m_reporters is a `Map<IReporter>` where r is `Class<? extends Reporter>`

Comment (juherr) [inline src/main/java/org/testng/TestNG.java]: You mean `Set<IReporter>`. But we can imagine was the intention here and I think it was not to remove the check ;)

File: src/main/java/org/testng/remote/strprotocol/GenericMessage.java (hunk 5 above)
Comment (juherr) [inline src/main/java/org/testng/remote/strprotocol/GenericMessage.java]: :+1:

File: src/main/java/org/testng/TestRunner.java (hunk 9 above)
Comment (testn) [inline src/main/java/org/testng/TestRunner.java]: m_verbose is already static before the code change.

File: src/main/java/org/testng/internal/MethodHelper.java (hunk 8 above)
Comment (testn) [inline src/main/java/org/testng/internal/MethodHelper.java]: will change

File: src/main/java/org/testng/TestRunner.java [hunk 11]
```diff
@@ -725,7 +724,7 @@ public int compareTo(IWorker<ITestNGMethod> other) {
       }
     });
 
-    runWorkers(workers, "" /* JUnit does not support parallel */, null);
+    runWorkers(workers);
```
Comment (testn) [inline src/main/java/org/testng/TestRunner.java]: this method has been refactored since it's only called by junit test. Do you think we should rename the method as well?

File: src/main/java/org/testng/reporters/JUnitReportReporter.java (hunk 4 above)
Comment (juherr) [inline src/main/java/org/testng/reporters/JUnitReportReporter.java]: Ok, but did you check what is the expected or used format from JUnit? 
It looks like `ISO 8601` is expected, so you don't need to imagine a custom pattern.

File: src/main/java/org/testng/TestRunner.java (hunk 11 above)
Comment (juherr) [inline src/main/java/org/testng/TestRunner.java]: Yes, I think "junit" should appear somewhere.

File: src/main/java/org/testng/TestRunner.java (hunk 9 above)
Comment (juherr) [inline src/main/java/org/testng/TestRunner.java]: Maybe, but I don't understand the goal to refactor something and keeping it wrong.

Comment (testn) [inline src/main/java/org/testng/TestRunner.java]: I don't intend to but if you search the usage.. there are 10+ places that use a static method TestRunner.getVerbose()

File: src/main/java/org/testng/internal/Utils.java (hunk 6 above)
Comment (testn) [inline src/main/java/org/testng/internal/Utils.java]: Will do

File: src/main/java/org/testng/reporters/JUnitReportReporter.java (hunk 4 above)
Comment (testn) [inline src/main/java/org/testng/reporters/JUnitReportReporter.java]: ISO 8601 will add "Z" instead of "GMT". If you are willing to accept this breaking change, I'm cool with that.

Comment (juherr) [inline src/main/java/org/testng/reporters/JUnitReportReporter.java]: I am but @cbeust will have the final word at the end. 
I'm just hoping he will found the time to have a look on it. 
The risk with too big pull request is to never be merged. That's why we asked little ones, easier to accept and merge.

File: src/main/java/org/testng/TestNG.java (hunk 10 above)
Comment (testn) [inline src/main/java/org/testng/TestNG.java]: By the way, `addReporter` is marked as private. The only place that uses it is in `initializeDefaultListeners`. For public interface, the consumer needs to do it via command line interface or use `addListener` which does not check any of this.

Comment (juherr) [inline src/main/java/org/testng/TestNG.java]: :+1:

File: src/main/java/org/testng/TestRunner.java (hunk 9 above)
Comment (juherr) [inline src/main/java/org/testng/TestRunner.java]: You choose. But this current change is wrong.

**Comment by juherr** [comment]
@testn Please stop to add new extra "improvments" in this PR which are not discuss before. I don't want to waste more time that I need.

File: src/main/java/org/testng/xml/XmlTest.java (hunk 2 above)
Comment (juherr) [inline src/main/java/org/testng/xml/XmlTest.java]: :+1:

File: src/main/java/org/testng/xml/XmlClass.java (hunk 3 above)
Comment (juherr) [inline src/main/java/org/testng/xml/XmlClass.java]: :+1:

File: src/main/java/org/testng/internal/Utils.java (hunk 6 above)
Comment (juherr) [inline src/main/java/org/testng/internal/Utils.java]: :+1:

File: src/main/java/org/testng/internal/TestNGClassFinder.java (hunk 7 above)
Comment (juherr) [inline src/main/java/org/testng/internal/TestNGClassFinder.java]: :+1:

File: src/main/java/org/testng/internal/MethodHelper.java (hunk 8 above)
Comment (juherr) [inline src/main/java/org/testng/internal/MethodHelper.java]: :+1:

File: src/main/java/org/testng/TestRunner.java (hunk 11 above)
Comment (juherr) [inline src/main/java/org/testng/TestRunner.java]: :+1:

File: src/main/java/org/testng/xml/XmlTest.java (hunk 1 above)
Comment (juherr) [inline src/main/java/org/testng/xml/XmlTest.java]: :+1:

File: src/main/java/org/testng/reporters/JUnitReportReporter.java (hunk 4 above)
Comment (juherr) [inline src/main/java/org/testng/reporters/JUnitReportReporter.java]: The goal of this reporter is to generated a JUnit XML report file. In fact, this report file was never generated by JUnit, but by the junit task of ant.
Have a look to [the usage of timestamp](https://fisheye6.atlassian.com/browse/~br=master/ant-git/src/main/org/apache/tools/ant/taskdefs/optional/junit/XMLJUnitResultFormatter.java?r=bf6026a147e879837eaad2a16348ae2162829867&r=df33426bf23dd97171dc3e0e7cb85de87d421bce#to138), and to [its expected format](https://fisheye6.atlassian.com/browse/~br=master/ant-git/src/main/org/apache/tools/ant/util/DateUtils.java?r=bf6026a147e879837eaad2a16348ae2162829867&r=df33426bf23dd97171dc3e0e7cb85de87d421bce&r=43844a7e6b54e2dd27216b30f6d7b44bb601d29c#to49).

**Comment by juherr** [comment]
@testn Could fix the 2 last issues (timestamp format and verbose)?

**Comment by testn** [comment]
- I cannot fix the verbose one as TestRunner.getVerbose() is mostly used with "static" context (i.e. it has no TestRunner object around
- For timestamp format, I already fixed it by using the date format only with no additional " GMT" string append.

File: src/main/java/org/testng/TestRunner.java (hunk 9 above)
Comment (juherr) [inline src/main/java/org/testng/TestRunner.java]: Please, revert this change.

File: src/main/java/org/testng/reporters/JUnitReportReporter.java (hunk 4 above)
Comment (juherr) [inline src/main/java/org/testng/reporters/JUnitReportReporter.java]: Please use "yyyy-MM-dd'T'HH:mm:ss" as format as it seems it what is used by ant.

**Comment by testn** [comment]
Let me know if you want me to address other issues.

**Comment by juherr** [comment]
I'm still disagree with the change of of timestamp (I provided you the format which shoud be used IMO) and the verbose (I prefer you revert this one).
@cbeust What do you think about this 2 changes? (I triple checked the other ones and I'm really confident).

**Comment by testn** [comment]
@juherr, I was just trying not to break the behavior of the existing code. What you proposed to change the toString() format. If you are ok with that, I am totally fine with it too.

**Comment by testn** [comment]
for verbose, changing the method to be static is only for correctness as the variable backing it is static so it does not really matter. I really wish that the code wasn't written that way to begin with. Now it is quite odd to tackle it after the fact.

**Comment by juherr** [comment]
* For `TestRunner.getVerbose()`, you are right from a Java point of view. But here we have some context we cannot ignore. And as you said, fix it is too much intrusive.

* For `toGMTString()`, I'm agree that not changing behavior is good. But I don't understand why we should keep something which seems wrong. I'd prefer @cbeust give us his point of view.

**Comment by cbeust** [comment]
This PR is already big enough so let's keep the old code in these two points as it was before.

We can discuss these two points in a separate issue or PR, it will be much easier than doing it here on this huge PR.

**Comment by juherr** [comment]
In others words:
* revert `getVerbose()`
* keep `toGMTString()` => #899

Allright!
@testn Could you finish it?

**Comment by juherr** [comment]
:+1:

**Comment by cbeust** [comment]
Thanks @testn !

**Comment by juherr** [comment]
Thanks @testn 

But clearly, I won't spend so much time to review another PR like that where the risk of regression is high with few value add. ;)

If you want to help, there are many issues to fix and test cases to add. Choose the one you want: https://github.com/cbeust/testng/issues :)

**Comment by cbeust** [comment]
Full agreement with @juherr.
